### PR TITLE
Use docker on travis

### PR DESCRIPTION
**Pull request by peter-edge**
This makes it so we no longer are using travis' env and now completely control the environment that fx is tested on. This also means that if it passes locally in docker, it passes remotely.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -1,28 +1,14 @@
...
-go_import_path: go.uber.org/fx
-
+sudo: required
+services:
+  - docker
 env:
-  - V=1 COVERMODE=atomic
-
+  - DOCKER_GO_VERSION=1.7 V=1 COVERMODE=atomic
+  - DOCKER_GO_VERSION=1.8 V=1 COVERMODE=atomic
```
Comment (anuptalwalkar) [inline .travis.yml]: Does this mean we can't run travis against `tip`?

File: Makefile [hunk 2]
```diff
@@ -14,7 +14,10 @@ all: lint test
 
 COV_REPORT := overalls.coverprofile
 
-DOCKER_IMAGE := uber/fx
+DOCKER_MAKE_CMD ?= test
```
Comment (anuptalwalkar) [inline Makefile]: Should this be part of `.build/docker.mk` ?

File: .travis.yml (hunk 1 above)
Comment (peter-edge) [inline .travis.yml]: Correct, there's no golang tip image

File: Makefile (hunk 2 above)
Comment (peter-edge) [inline Makefile]: I actually dislike having separate build files like this, it makes things less explicit, I think it's fine here, not much logic and it directly links to logic in the Makefile

File: .travis.yml (hunk 1 above)
Comment (peter-edge) [inline .travis.yml]: Personally this not only doesn't bother me, but I rather not be needing to test against tip - it's not like the one time every 6 months that golang is released we can't re-run `make dockerrun` :)

File: Makefile (hunk 2 above)
Comment (glibsm) [inline Makefile]: `make ci` here, right?

Comment (peter-edge) [inline Makefile]: Really I would not want any default tbh...actually I'm going to push something

File: Makefile [hunk 3]
```diff
@@ -128,15 +131,11 @@ benchreset:
 
 .PHONY: dockerbuild
 dockerbuild:
-	docker build -t $(DOCKER_IMAGE) .
+	docker build -t $(DOCKER_IMAGE) -f $(DOCKERFILE) .
```
Comment (breerly) [inline Makefile]: Why not build *inside* of the Docker container?

File: Makefile [hunk 4]
```diff
@@ -14,7 +14,10 @@ all: lint test
 
 COV_REPORT := overalls.coverprofile
 
-DOCKER_IMAGE := uber/fx
+DOCKER_MAKE_CMD ?= test
+DOCKER_GO_VERSION ?= 1.8
+DOCKER_IMAGE := uber/fx-$(DOCKER_GO_VERSION)
```
Comment (breerly) [inline Makefile]: Why not name the image the same thing each build?

**Comment by breerly** [comment]
Totally in favor of this - reproducible builds is something we should lean on Mr. Docker for.

File: Makefile (hunk 4 above)
Comment (peter-edge) [inline Makefile]: Cuz once I get to it, I will add caching http://rundef.com/fast-travis-ci-docker-build

File: Makefile (hunk 3 above)
Comment (peter-edge) [inline Makefile]: Wait what docker container haha.

**Comment by peter-edge** [comment]
@breerly I can set this up for zap, sally, etc too...this is what I generally like to do sometimes anyways, especially for non-trivial stuff https://github.com/peter-edge/protoeasy-go/blob/master/circle.yml

**Comment by glibsm** [comment]
Im ok to give up tip and see if that significantly improves our build times. Always easy to put it back.

**Comment by peter-edge** [comment]
This PR is going to slow them down temporarily. Setting up caching is not fun. But it still is an improvement IMO.

**Comment by peter-edge** [comment]
Slow down = from 2 mins to 3 mins

**Pull request merged**